### PR TITLE
Topology support

### PR DESCRIPTION
**Pull request by verult**
Creating a PR for early review.

I have a local working prototype. The hope is each of these PRs introduce a feature whose implementation is as close to the final form as possible.

@jsafrane @msau42 @saad-ali @ddebroy

Features (some of these need more refinement but core logic is there):
- [x] Generating PV Node Affinity
- [x] Generating CSI requisite topology: Parsing AllowedTopologies
- [x] Generating CSI requisite topology: Topology aggregation across the cluster, when AllowedTopologies is missing
- [x] Generating CSI preferred topology
- [ ] Evenly spreading topology preference for PVs in StatefulSets.

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -172,10 +177,27 @@ func supportsPluginControllerService(conn *grpc.ClientConn, timeout time.Duratio
...
+	rsp, err := client.ControllerGetCapabilities(ctx, &req)
+	if err != nil {
+		return nil, err
 	}
-	caps := rsp.GetCapabilities()
-	for _, cap := range caps {
+	return rsp.GetCapabilities(), nil
+}
+
+func supportsPluginControllerService(capabilities []*csi.PluginCapability) bool {
```
Comment (msau42) [inline pkg/controller/controller.go]: You could have one helper method that searches capabilities for the passed in service

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -268,27 +284,27 @@ func NewCSIProvisioner(client kubernetes.Interface,
...
 // Before initiating Create/Delete API calls provisioner checks if Capabilities:
 // PluginControllerService,  ControllerCreateVolume sre supported and gets the  driver name.
-func checkDriverState(grpcClient *grpc.ClientConn, timeout time.Duration, needSnapshotSupport bool) (string, error) {
-	ok, err := supportsPluginControllerService(grpcClient, timeout)
+func checkDriverState(grpcClient *grpc.ClientConn, timeout time.Duration, needSnapshotSupport bool) (*driverState, error) {
+	pluginCaps, err := getPluginCapabilities(grpcClient, timeout)
 	if err != nil {
-		glog.Errorf("failed to get support info :%v", err)
-		return "", err
+		glog.Errorf("failed to get plugin capabilities: %v", err)
```
Comment (msau42) [inline pkg/controller/controller.go]: This should be handled as a separate pr, but I want to get rid of this style of logging errors that we're also returning

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -90,6 +90,11 @@ type csiProvisioner struct {
 	config               *rest.Config
 }
 
+type driverState struct {
+	driverName string
+	accessibilityConstraintCapability bool
```
Comment (msau42) [inline pkg/controller/controller.go]: a more clear name could be "supportsAccessiblity"

Comment (msau42) [reply pkg/controller/controller.go]: or you could potentially make this generic and have an enum of all the different capabilities

File: pkg/controller/controller.go (hunk 2 above)
Comment (verult) [reply pkg/controller/controller.go]: I'll address it in this PR. Improving the code base one line at a time :)

**Comment by verult** [comment]
All the core functionality and the vast majority of core testing is complete. Reviewers PTAL.

I'll add StatefulSet spreading in a separate PR post k8s 1.12 release since it's an optimization.

Working on the TODOs left in the PR but I'll hold off on pushing mode code for now until we get a critical mass of reviews
